File: ui/src/decks/Deck.ts
import { Deck } from "@material-ui/icons"
import { HasAerc } from "../aerc/HasAerc"
import { KCard } from "../cards/KCard"
import { log, roundToHundreds } from "../config/Utils"
import { BackendExpansion } from "../expansions/Expansions"
import { House } from "../houses/House"
import { DeckSynergyInfo } from "../synergy/DeckSynergyInfo"
import { userStore } from "../user/UserStore"
import { UserDeck } from "../userdeck/UserDeck"
import { userDeckStore } from "../userdeck/UserDeckStore"
import { DeckSaleInfo } from "./sales/DeckSaleInfo"

export interface DeckWithSynergyInfo {
    deck: Deck
    cardRatingPercentile: number
    synergyPercentile: number
    antisynergyPercentile: number
}

export interface Deck extends HasAerc {

    id: number
    keyforgeId: string
    name: string
    expansion: BackendExpansion
    powerLevel: number
    chains: number
    wins: number
    losses: number
    crucibleTrackerWins?: number
    crucibleTrackerLosses?: number

    registered: boolean

    maverickCount: number
    specialsCount: number
    raresCount: number
    uncommonsCount: number

    rawAmber: number
    totalPower: number
    creatureCount: number
    actionCount: number
    artifactCount: number
    upgradeCount: number

    cardDrawCount: number
    cardArchiveCount: number
    keyCheatCount: number
    totalArmor: number

    aercScore: number
    previousSasRating: number
    sasV3?: number
    sasRating: number
    synergyRating: number
    antisynergyRating: number

    forSale: boolean
    forTrade: boolean
    forAuction: boolean
    wishlistCount: number
    funnyCount: number

    lastSasUpdate: string

    sasPercentile?: number

    userDecks: UserDeck[]

    houses: House[]

    deckSaleInfo?: DeckSaleInfo[]

    owners?: string[]

    synergies?: DeckSynergyInfo

}

export class DeckUtils {

    static findPrice = (deck: Deck, myPriceOnly?: boolean): number | undefined => {
        const saleInfo = deck.deckSaleInfo
        if (saleInfo && saleInfo.length > 0) {
            for (const info of saleInfo) {
                if (!myPriceOnly || info.username === userStore.username) {
                    return info.buyItNow
                }
            }
        }
        return undefined
    }

    static findHighestBid = (deck: Deck): number | undefined => {
        const saleInfo = deck.deckSaleInfo
        if (saleInfo && saleInfo.length > 0) {
            for (const info of saleInfo) {
                if (info.highestBid != null) {
                    return info.highestBid
                }
            }
        }
        return undefined
    }

    static hasAercFromDeck = (deck: Deck): HasAerc => {
        if (deck.synergies == null) {
            log.warn("Synergies shouldnt' be null!")
            return deck
        }
        return {
            aercScore: deck.synergies.rawAerc,
            ...deck.synergies
        }
    }

    static cardsInHouses = (deck: Deck) => {
        const cardsByHouse: { house: House, cards: KCard[] }[] = []
        deck.houses
            .forEach((house) => {
                const cards = deck.searchResultCards!.filter(card => card.house === house)
                cardsByHouse.push({house, cards})
            })
        return cardsByHouse
    }

    static synergiesRounded = (synergies: DeckSynergyInfo) => {
        const {
            amberControl,
            expectedAmber,
            amberProtection,
            artifactControl,
            creatureControl,
            effectivePower,
            efficiency,
            disruption,
            houseCheating,
            other,
            ...rest
        } = synergies
        return {
            amberControl: roundToHundreds(amberControl),
            expectedAmber: roundToHundreds(expectedAmber),
            amberProtection: roundToHundreds(amberProtection),
            artifactControl: roundToHundreds(artifactControl),
            creatureControl: roundToHundreds(creatureControl),
            effectivePower: roundToHundreds(effectivePower),
            efficiency: roundToHundreds(efficiency),
            disruption: roundToHundreds(disruption),
            houseCheating: roundToHundreds(houseCheating),
            other: roundToHundreds(other),
            ...rest
        }
    }

    static arrayToCSV = (decks: Deck[]) => {
        const data = decks.map(deck => {
            const synergies = DeckUtils.synergiesRounded(deck.synergies!)
            return [
                deck.name.replace(/"/g, "\"\""),
                deck.houses,
                deck.expansion,
                synergies.sasRating,
                synergies.synergyRating,
                synergies.antisynergyRating,
                deck.sasPercentile,
                synergies.rawAerc,
                synergies.amberControl,
                synergies.expectedAmber,
                synergies.amberProtection,
                synergies.artifactControl,
                synergies.creatureControl,
                synergies.effectivePower,
                synergies.efficiency,
                synergies.disruption,
                synergies.houseCheating,
                synergies.other,

                deck.creatureCount,
                deck.actionCount,
                deck.artifactCount,
                deck.upgradeCount,

                deck.rawAmber,
                deck.keyCheatCount,
                deck.cardDrawCount,
                deck.cardArchiveCount,
                deck.totalPower,
                deck.totalArmor,

                deck.powerLevel,
                deck.chains,
                deck.wins,
                deck.losses,

                deck.forSale,
                deck.forAuction,
                deck.forTrade,
                DeckUtils.findPrice(deck),
                deck.searchResultCards == null ? "" : `${deck.searchResultCards.map(card => card.cardTitle).join("|")}`,
                deck.wishlistCount,
                deck.funnyCount,
                `https://decksofkeyforge.com/decks/${deck.keyforgeId}`,
                `https://www.keyforgegame.com/deck-details/${deck.keyforgeId}`,
                deck.lastSasUpdate,

                userDeckStore.notesForDeck(deck.id) ?? ""
            ]
        })
        data.unshift([
            "Name",
            "Houses",
            "Expansion",
            "Sas Rating",
            "Synergy Rating",
            "Antisynergy Rating",
            "Sas Percentile",
            "Raw Aerc Score",
            "Amber Control",
            "Expected Amber",
            "Aember Protection",
            "Artifact Control",
            "Creature Control",
            "Effective Power",
            "Efficiency",
            "Disruption",
            "House Cheating",
            "Other",

            "Creature Count",
            "Action Count",
            "Artifact Count",
            "Upgrade Count",

            "Raw Amber",
            "Key Cheat Count",
            "Card Draw Count",
            "Card Archive Count",
            "Total Power",
            "Total Armor",

            "Power Level",
            "Chains",
            "Wins",
            "Losses",

            "For Sale",
            "For Auction",
            "For Trade",
            "Price",

            "Cards",
            "Wishlist",
            "Funny",
            "DoK Link",
            "Master Vault Link",
            "Last SAS Update",

            "My Notes"
        ])
        return data
    }
}

export interface DeckPage {
    decks: Deck[]
    page: number
}

export interface DeckCount {
    pages: number
    count: number
}

File: src/main/kotlin/coraythan/keyswap/decks/models/Deck.kt
package coraythan.keyswap.decks.models

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import coraythan.keyswap.*
import coraythan.keyswap.auctions.DeckListing
import coraythan.keyswap.cards.CardType
import coraythan.keyswap.cards.dokcards.DokCardInDeck
import coraythan.keyswap.cards.dokcards.toUrlFriendlyCardTitle
import coraythan.keyswap.keyforgeevents.tournaments.TournamentDeck
import coraythan.keyswap.stats.DeckStatistics
import coraythan.keyswap.synergy.DeckSynergyInfo
import coraythan.keyswap.synergy.SynTraitPlayer
import coraythan.keyswap.synergy.SynergyTrait
import coraythan.keyswap.synergy.containsTrait
import coraythan.keyswap.tags.DeckTag
import coraythan.keyswap.thirdpartyservices.mastervault.KeyForgeDeck
import coraythan.keyswap.userdeck.*
import jakarta.persistence.*
import java.time.LocalDate
import java.time.ZonedDateTime

@Entity
data class Deck(

    @Column(unique = true)
    val keyforgeId: String,

    override val name: String,
    override val expansion: Int,
    val powerLevel: Int = 0,
    val chains: Int = 0,
    val wins: Int = 0,
    val losses: Int = 0,

    val previousSasRating: Int? = 0,
    val previousMajorSasRating: Int? = 0,

    val forSale: Boolean = false,
    val forTrade: Boolean = false,
    val forAuction: Boolean = false,
    val completedAuction: Boolean = false,
    val wishlistCount: Int = 0,
    val funnyCount: Int = 0,
    val evilTwin: Boolean? = false,

    override val cardIds: String = "",
    override val tokenNumber: Int? = null,

    override val houseNamesString: String = "",

    override val bonusIconsString: String? = null,

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val deckNotes: List<DeckNote> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val tournamentDecks: List<TournamentDeck> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val ownedDecks: List<OwnedDeck> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val funnyDecks: List<FunnyDeck> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val favoritedDecks: List<FavoritedDeck> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val previouslyOwnedDecks: List<PreviouslyOwnedDeck> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    val tags: List<DeckTag> = listOf(),

    @JsonIgnoreProperties("deck")
    @OneToMany(mappedBy = "deck", fetch = FetchType.LAZY)
    val auctions: List<DeckListing> = listOf(),

    val hasOwnershipVerification: Boolean? = false,

    val listedOn: ZonedDateTime? = null,
    val auctionEnd: ZonedDateTime? = null,
    val auctionEndedOn: ZonedDateTime? = null,

    val importDateTime: ZonedDateTime? = now(),

    /**
     * Last SAS update
     */
    val lastUpdate: ZonedDateTime? = now(),

    val twinId: String? = null,

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO, generator = "hibernate_sequence")
    @SequenceGenerator(name = "hibernate_sequence", allocationSize = 1)
    val id: Long = -1
) : GenericDeck {

    val dateAdded: LocalDate?
        get() = this.importDateTime?.toLocalDate()

    fun withBonusIcons(icons: DeckBonusIcons): Deck {
        return this
            .copy(bonusIconsString = jacksonObjectMapper().writeValueAsString(icons))
    }

    override fun toDeckSearchResult(
        housesAndCards: List<HouseAndCards>?,
        cards: List<DokCardInDeck>?,
        stats: DeckStatistics?,
        synergies: DeckSynergyInfo?,
        includeDetails: Boolean,
        token: DokCardInDeck?,
    ): DeckSearchResult {

        return DeckSearchResult(
            deckType = DeckType.STANDARD,
            id = id,
            keyforgeId = keyforgeId,
            expansion = expansionEnum,
            name = name,

            powerLevel = powerLevel.zeroToNull(),
            chains = chains.zeroToNull(),
            wins = wins.zeroToNull(),
            losses = losses.zeroToNull(),

            creatureCount = cards?.count { it.card.cardType == CardType.Creature },
            actionCount = cards?.count { it.card.cardType == CardType.Action },
            artifactCount = cards?.count { it.card.cardType == CardType.Artifact },
            upgradeCount = cards?.count { it.card.cardType == CardType.Upgrade },

            cardDrawCount = (cards?.filter {
                it.extraCardInfo.traits.containsTrait(SynergyTrait.drawsCards) || it.extraCardInfo.traits.containsTrait(
                    SynergyTrait.increasesHandSize
                )
            }?.size ?: 0).zeroToNull(),
            cardArchiveCount = (cards?.filter {
                it.extraCardInfo.traits.containsTrait(
                    SynergyTrait.archives,
                    player = SynTraitPlayer.FRIENDLY
                )
            }?.size
                ?: 0).zeroToNull(),
            keyCheatCount = (cards?.filter {
                it.extraCardInfo.traits.containsTrait(SynergyTrait.forgesKeys)
                        || it.extraCardInfo.traits.containsTrait(SynergyTrait.forgesKeysWithoutAember)
            }?.size ?: 0).zeroToNull(),
            rawAmber = if (cards == null) -1 else cards.sumOf { it.card.amber } + this.bonusIcons().bonusIconHouses.sumOf { it.bonusIconCards.sumOf { it.bonusAember } },
            totalArmor = cards?.sumOf { it.card.armor },
            totalPower = cards?.sumOf { it.card.power } ?: -1,

            expectedAmber = synergies?.expectedAmber ?: -1.0,
            amberControl = synergies?.amberControl ?: -1.0,
            creatureControl = synergies?.creatureControl ?: -1.0,
            artifactControl = (synergies?.artifactControl ?: -1.0).zeroToNull(),
            efficiency = (synergies?.efficiency ?: -1.0).zeroToNull(),
            recursion = (synergies?.recursion ?: -1.0).zeroToNull(),
            effectivePower = synergies?.effectivePower ?: -1,
            creatureProtection = (synergies?.creatureProtection ?: -1.0).zeroToNull(),
            disruption = (synergies?.disruption ?: -1.0).zeroToNull(),
            other = (synergies?.other ?: -1.0).zeroToNull(),
            aercScore = synergies?.rawAerc ?: -1,
            previousSasRating = previousSasRating ?: -1,
            previousMajorSasRating = previousMajorSasRating,
            aercVersion = -1,
            sasRating = synergies?.sasRating ?: -1,
            synergyRating = synergies?.synergyRating ?: -1,
            antisynergyRating = synergies?.antisynergyRating ?: -1,
            forSale = forSale.falseToNull(),
            forTrade = forTrade.falseToNull(),
            forAuction = forAuction.falseToNull(),
            wishlistCount = wishlistCount.zeroToNull(),
            funnyCount = funnyCount.zeroToNull(),
            housesAndCards = housesAndCards?.addBonusIcons(bonusIcons()) ?: listOf(),

            lastSasUpdate = lastUpdate?.toLocalDateWithOffsetMinutes(-420)?.toString().emptyToNull(),

            sasPercentile = stats?.sasStats?.closestPercentileForValue(synergies?.sasRating ?: 0) ?: 0.0,

            synergyDetails = if (includeDetails) synergies?.synergyCombos else synergies?.synergyCombos?.map {
                it.copy(
                    synergies = listOf()
                )
            },
            tokenCreationValues = synergies?.tokenCreationValues,
            efficiencyBonus = synergies?.efficiencyBonus ?: 0.0,

            hasOwnershipVerification = hasOwnershipVerification.falseToNull(),

            dateAdded = dateAdded,

            twinId = twinId,
            tokenInfo = token?.toTokenInfo(),
            hauntingOdds = synergies?.hauntingOdds,
        )
    }

    fun addGameStats(keyforgeDeck: KeyForgeDeck): Deck? {
        if (this.wins == keyforgeDeck.wins && this.losses == keyforgeDeck.losses
            && this.chains == keyforgeDeck.chains && this.powerLevel == keyforgeDeck.power_level
        ) {
            return null
        }
        return this.copy(
            wins = keyforgeDeck.wins,
            losses = keyforgeDeck.losses,
            chains = keyforgeDeck.chains,
            powerLevel = keyforgeDeck.power_level
        )
    }

}

fun List<DokCardInDeck>.genCardNamesIndexableString(): String {
    return "~" +
            // Add the cards themselves
            this
                .groupBy { it.card.cardTitle }
                .map { entry ->
                    "${entry.key}${(1..entry.value.size).joinToString("")}"
                }.sorted().joinToString("~") + "~" +
            // Add duplicates for mavericks with the house
            this
                .asSequence()
                .filter { it.maverick && !it.anomaly }
                .groupBy { it.card.cardTitle }
                .flatMap { entry ->
                    entry.value
                        .groupBy { it.house }
                        .map { houseToCards ->
                            val firstCard = houseToCards.value[0]
                            "${firstCard.card.cardTitle}${firstCard.house}"
                        }
                }.sorted().joinToString("~") + "~" +
            // Add duplicates for anomalies with the house
            this
                .asSequence()
                .filter { it.anomaly }
                .groupBy { it.card.cardTitle }
                .flatMap { entry ->
                    entry.value
                        .groupBy { it.house }
                        .map { houseToCards ->
                            val firstCard = houseToCards.value[0]
                            "${firstCard.card.cardTitle}${firstCard.house}"
                        }
                }.sorted().joinToString("~") + "~"
}

@JsonIgnoreProperties(ignoreUnknown = true)
data class DecksPage(
    val decks: List<DeckSearchResult>,
    val page: Long
)

data class DeckCount(
    val pages: Long,
    val count: Long
)

data class DeckBonusIcons(
    val bonusIconHouses: List<BonusIconHouse> = listOf()
)

data class BonusIconHouse(
    val house: House,
    val bonusIconCards: List<BonusIconsCard> = listOf(),
)

data class BonusIconsCard(
    val cardTitle: String,
    val bonusAember: Int = 0,
    val bonusCapture: Int = 0,
    val bonusDamage: Int = 0,
    val bonusDraw: Int = 0,
    val bonusDiscard: Int = 0,
)

fun List<DokCardInDeck>.withBonusIcons(icons: DeckBonusIcons): List<DokCardInDeck> {
    if (icons.bonusIconHouses.isEmpty()) return this
    return this.groupBy { it.house }
        .map { houseAndCards ->
            val bonusIconsCards =
                icons.bonusIconHouses.first { it.house == houseAndCards.key }.bonusIconCards.toMutableList()
            houseAndCards.value.map { dokCardInDeck ->
                val bonusIcons =
                    bonusIconsCards.find { cardIcons -> cardIcons.cardTitle.toUrlFriendlyCardTitle() == dokCardInDeck.card.cardTitle.toUrlFriendlyCardTitle() }
                bonusIconsCards.remove(bonusIcons)
                dokCardInDeck.copy(
                    bonusAember = bonusIcons?.bonusAember ?: 0,
                    bonusCapture = bonusIcons?.bonusCapture ?: 0,
                    bonusDamage = bonusIcons?.bonusDamage ?: 0,
                    bonusDraw = bonusIcons?.bonusDraw ?: 0,
                    bonusDiscard = bonusIcons?.bonusDiscard ?: 0,
                )
            }
        }
        .flatten()
}
